examples/active_replication/flux_producer_operator.py

```diff
@@ -0,0 +1,93 @@
+from erdos.data_stream import DataStream
+from erdos.op import Op
+from erdos.utils import setup_logging
+
+import flux_utils
+from flux_utils import is_ack_stream, is_not_ack_stream, is_not_back_pressure
+from flux_buffer import Buffer
+
+import threading
+
+
+class FluxProducerOperator(Op):
+    def __init__(self,
+                 name,
+                 replica_num,
+                 output_stream_name,
+                 log_file_name=None):
+        super(FluxProducerOperator, self).__init__(name)
+        self._logger = setup_logging(self.name, log_file_name)
+        self._replica_num = replica_num
+        self._ex_secondary = False
+        self._output_stream_name = output_stream_name
+        self._output_seq_num = 0
+        if replica_num > 0:
+            # only replicas have these buffers to keep track
+            # with the progress of primary op
+            self._buffer = Buffer(1)
+            self._ack_buffer = set()
+        self.lock = threading.Lock()
+
+    @staticmethod
+    def setup_streams(input_streams, output_stream_name):
+        input_streams\
+            .filter(is_not_ack_stream)\
+            .filter(is_not_back_pressure)\
+            .add_callback(FluxProducerOperator.on_msg)
+        input_streams.filter(is_ack_stream).add_callback(
+            FluxProducerOperator.on_ack_msg)
+
+        return [DataStream(name=output_stream_name)]
+
+    def on_msg(self, msg):
+        self.lock.acquire()
+        # XXX(ionel): Need to ensure that we remove output_seq_num
+        # during recovery.
+        if self._replica_num == 0:  # Primary
+            msg.data = (self._output_seq_num, msg.data)
+            if self._ex_secondary:  # used to a secondary
```

Comment:
Will this code ever get invoked because it is in the self._replica_num == 0 body. I think it's safer to have another attribute self._master which you can set. In this way, you won't make the assumption that the master always has replica_num = 0.

Reply:
Also, might be worthwhile to push all this recovery code into a separate method.

Reply:
I feel like self._replica_num is completely equivalent to self._master.

Reply:
Imagine you have a primary (replica_num == 0), and a secondary (replica_num == 1). If the primary fails and the secondary becomes primary then the primary will be the node with replica_num == 1. I don't think they're equivalent.

Yes, you do set the replica_num to 0 when you detect that the primary has failed, but IMHO it's easy to introduce bugs if you do that. What if any buffer uses replica num (id), or if any upstream/downstream operator works with replica id. It's just dangerous to change replica id in flight.

Reply:
Ok will change it. But it will just be renaming self._replica_num to self._primary because I only use self._replica_num for this purpose.

Reply:
Yep, that should do it, if you rename it to self._primary and make it a boolean.